### ISSUE TITLE
CUDA_ERROR_ILLEGAL_ADDRESS: "an illegal memory access was encountered" with candle_nn `Silu`

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#1830]
Hello all,

I have come across a CUDA_ERROR_ILLEGAL_ADDRESS error, with the traceback below. This occurs when `silu` is run. I do not implement parallelism for the model itself, and it runs with a single thread. 

I have seen #1385, which does not have an answer. @LaurentMazare, how can I check if the tensor has been dropped (it appears there is no way to check the internal refcount of the `Arc`), or is there something else possibly going wrong? The error occurs in the `Drop` impl of `Arc`, which seems to me like a dangling reference was created. I would very much appreciate any help to fix this.

```bash
thread '<unnamed>' panicked at /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cudarc-0.10.0/src/driver/safe/core.rs:208:76:
called `Result::unwrap()` on an `Err` value: DriverError(CUDA_ERROR_ILLEGAL_ADDRESS, "an illegal memory access was encountered")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/cudarc-0.10.0/src/driver/safe/core.rs:208:76:
called `Result::unwrap()` on an `Err` value: DriverError(CUDA_ERROR_ILLEGAL_ADDRESS, "an illegal memory access was encountered")
stack backtrace:
   0:     0x5620fca9acc6 - std::backtrace_rs::backtrace::libunwind::trace::hbee8a7973eeb6c93
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x5620fca9acc6 - std::backtrace_rs::backtrace::trace_unsynchronized::hc8ac75eea3aa6899
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x5620fca9acc6 - std::sys_common::backtrace::_print_fmt::hc7f3e3b5298b1083
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x5620fca9acc6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbb235daedd7c6190
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x5620fcacb430 - core::fmt::rt::Argument::fmt::h76c38a80d925a410
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/fmt/rt.rs:142:9
   5:     0x5620fcacb430 - core::fmt::write::h3ed6aeaa977c8e45
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/fmt/mod.rs:1120:17
   6:     0x5620fca9744f - std::io::Write::write_fmt::h78b18af5775fedb5
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/io/mod.rs:1810:15
   7:     0x5620fca9aaa4 - std::sys_common::backtrace::_print::h5d645a07e0fcfdbb
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x5620fca9aaa4 - std::sys_common::backtrace::print::h85035a511aafe7a8
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x5620fca9c327 - std::panicking::default_hook::{{closure}}::hcce8cea212785a25
  10:     0x5620fca9c089 - std::panicking::default_hook::hf5fcb0f213fe709a
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:292:9
  11:     0x5620fca9c7b8 - std::panicking::rust_panic_with_hook::h095fccf1dc9379ee
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:779:13
  12:     0x5620fca9c692 - std::panicking::begin_panic_handler::{{closure}}::h032ba12139b353db
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:657:13
  13:     0x5620fca9b1c6 - std::sys_common::backtrace::__rust_end_short_backtrace::h9259bc2ff8fd0f76
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys_common/backtrace.rs:171:18
  14:     0x5620fca9c3f0 - rust_begin_unwind
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
  15:     0x5620fc114545 - core::panicking::panic_fmt::h784f20a50eaab275
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
  16:     0x5620fc114a93 - core::result::unwrap_failed::h03d8a5018196e1cd
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/result.rs:1649:5
  17:     0x5620fc2eef14 - <cudarc::driver::safe::core::CudaSlice<T> as core::ops::drop::Drop>::drop::h346fb06c9ea331ee
  18:     0x5620fc2fabd4 - alloc::sync::Arc<T,A>::drop_slow::h0ba9570eef3efe07
  19:     0x5620fc2fb780 - alloc::sync::Arc<T,A>::drop_slow::h967574ea0c28f711
  20:     0x5620fc22283c - <mistralrs_core::models::quantized_llama::Mlp as candle_core::Module>::forward::h86daf01e771f1fb0
  21:     0x5620fc222b0d - <mistralrs_core::models::quantized_llama::MlpOrMoe as candle_core::Module>::forward::heeb5f517c1c2e137
  22:     0x5620fc239d16 - mistralrs_core::models::quantized_llama::ModelWeights::forward::h099dfc92cebfb06d
  23:     0x5620fc28f151 - <mistralrs_core::pipeline::mistral::MistralPipeline as mistralrs_core::pipeline::Pipeline>::forward::h13f21c2f9558554a
  24:     0x5620fc2008f7 - mistralrs_core::engine::Engine::run::h621f9a92d3156163
  25:     0x5620fc1f50d3 - std::sys_common::backtrace::__rust_begin_short_backtrace::hdf757670f422fcf5
  26:     0x5620fc2b615c - core::ops::function::FnOnce::call_once{{vtable.shim}}::h0b1fd19c1138ec5f
  27:     0x5620fcaa1fc5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h12de4fc57affb195
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/alloc/src/boxed.rs:2015:9
  28:     0x5620fcaa1fc5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h3c619f45059d5cf1
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/alloc/src/boxed.rs:2015:9
  29:     0x5620fcaa1fc5 - std::sys::unix::thread::Thread::new::thread_start::hbac657605e4b7389
                               at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/sys/unix/thread.rs:108:17
  30:     0x7fd040694ac3 - start_thread
                               at ./nptl/pthread_create.c:442:8
  31:     0x7fd040726a40 - clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  32:                0x0 - <unknown>
thread '<unnamed>' panicked at library/core/src/panicking.rs:163:5:
panic in a destructor during cleanup
thread caused non-unwinding panic. aborting.
Aborted (core dumped)
```

**Comment by LaurentMazare** [comment]
Could you try to make a small repro for this, ideally something without any dependency that is only a couple of lines long so that it's easy to replicate the issue locally.
Looking at the backtrace, it doesn't seem to be about a tensor being dropped but rather a quantized layer (or qtensor) which is fairly different from the other issue. We don't do anything specific in the drop part of these so pretty unclear what could cause this. You could also try running with `CUDA_LAUNCH_BLOCKING=1` just in case to see if it's not that the error occured at another location and only gets reported here. You can also try the cuda compute sanitizer to pinpoint at a more precise place though it's a bit of an advanced tool.

**Comment by EricLBuehler** [comment]
Thank you for the Running with `CUDA_LAUNCH_BLOCKING=1` failed with the same traceback. I am interested in trying the compute sanitizer, but how can I use it if I cannot control the `nvcc` invocation? Additionally, the following script does not cause an error:

```rust
fn tester() {
    use candle_core::{Device, DType, Tensor, Module};
    let dev = Device::new_cuda(0).unwrap();
    let q = candle_core::quantized::QTensor::quantize(
        &Tensor::zeros((4096, 14336), DType::F32, &dev).unwrap(),
        candle_core::quantized::GgmlDType::Q4_0,
    ).unwrap();
    let matmul = candle_core::quantized::QMatMul::from_qtensor(q).unwrap();
    
    let inp = Tensor::zeros((2,4096,14336), DType::F32, &dev).unwrap();
    dbg!(&inp);

    let res = matmul.forward(&inp).unwrap();
    dbg!(&res);

    let silued = candle_nn::ops::silu(&res).unwrap();
    dbg!(silued);
}
```

I think this simulates the situation: The model is set up (and the QMatMul is set up), and then when the Mlp layer is run it runs the QMatMul. The `silu` layer is where the error occurs in the real program. However, clearly there is something missing, as it does not match the error.

I ran the `quantized` example and noticed that it is using 86 GB of virtual memory and `mistralrs-server` is also using 86 GB of virtual memory and 5 GB of VRAM. This seems very strange, and perhaps it is connected. Do you have any thoughts?

**Comment by LaurentMazare** [comment]
You can use compute-sanitizer without tweking the `nvcc` flags though it would give you less insights about the actual lines where things happen. If this is not enough there are environment variables that can be used to tweak the `nvcc` flags without having to change the way it's called.
Re virtual memory usage, I think this is expected when one run some cuda program, I think it maps the whole host memory + gpu memory so as to have some unified memory addressing, this doesn't mean that it's actually using the memory.

**Comment by EricLBuehler** [comment]
Ok, thank you for the clarification about virtual memory usage. I will research it a bit more. 

I have created a minimum reproducible example on top of the Candle Quantized example, [here](https://github.com/EricLBuehler/candle/blob/segfault/candle-examples/examples/quantized/main.rs). The only lines I changed are [these 3](https://github.com/EricLBuehler/candle/blob/3184030de6d4d7db91637e52e6a6e50beb9e1152/candle-examples/examples/quantized/main.rs#L499-L501), and just add a second batch (the start offset can stay the same because the prompt is the same, although in `mistral.rs` we of course do something more intelligent).

It looks like the second batch exposes/causes this bug somehow, I hope this is enough for you to take a look?
